### PR TITLE
Refactor the HttplugExtension

### LINKED ISSUE #95: Improve HttplugExtension
| Q            | A
| ------------ | ---
| Bug?         | yes
| New Feature? | no
| Version      | 1.2.x

#### Actual Behavior

The extension became quite unstable with the 1.2.0 release. Profiling toolbar is not so well separated (enabled in production mode, etc). This has to be improved.

### PR DESCRIPTION
**Pull request by Nyholm**
| Q               | A
| --------------- | ---
| Bug fix?        | yes, or maybe..
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | yes
| Related tickets | #95
| Documentation   | n/a
| License         | MIT

#### What's in this PR?

Since the bugs discovered in our extension class I started to refactor it so it would be easier to maintain and understand. This PR will fix #95 or at least be a start of it. 

Major changes:

- All clients are registered with the PluginClient. (even in production)
- `auto` toolbar option has been deprecated, defaults to kernel debug

### CODE REVIEW

**Comment by dbu** [comment]
cool, thanks! https://github.com/php-http/HttplugBundle/blob/master/Tests/Unit/DependencyInjection/HttplugExtensionTest.php is not a comprehensive test, but it looks like at least the basics have no changes. 

i think this is going in the right direction. are you confident with it, should we merge it?

**Comment by Nyholm** [comment]
Im confident, but.. I was also confident with 1.2.0. If it works when you try it Im happy to merge. More PRs will come up with more test code.

**Comment by Nyholm** [comment]
I've rebased this PR so we include https://github.com/php-http/HttplugBundle/pull/98. That would make me more confident.

**Comment by dbu** [comment]
can you rebase so that we see if the new tests succeed?

**Comment by dbu** [comment]
ups, faster than me :-)

@sagikazarmark LGTM for me, can you have a look too?

File: ClientFactory/PluginClientFactory.php [hunk 1]
```diff
@@ -13,15 +13,23 @@
...
-    public static function createPluginClient(array $plugins, ClientFactory $factory, array $config, array $pluginClientOptions = [])
+    public static function createPluginClient(array $plugins, $factory, array $config, array $pluginClientOptions = [])
     {
-        return new PluginClient($factory->createClient($config), $plugins, $pluginClientOptions);
+        if ($factory instanceof ClientFactory) {
+            $client = $factory->createClient($config);
+        } elseif (is_callable($factory)) {
+            $client = $factory($config);
+        } else {
+            throw new \RuntimeException(sprintf('Second argument to PluginClientFactory::createPluginClient must be a "%s" or a callale.', ClientFactory::class));
```
Comment (joelwurtz) [inline ClientFactory/PluginClientFactory.php]: typo: callale -> callable

File: DependencyInjection/HttplugExtension.php [hunk 2]
```diff
@@ -66,45 +50,47 @@ public function load(array $configs, ContainerBuilder $container)
         }
 
         $this->configurePlugins($container, $config['plugins']);
-        $this->configureClients($container, $config, $toolbar);
-        $this->configureAutoDiscoveryClients($container, $config, $toolbar);
+        $serviceIds = $this->configureClients($container, $config);
+        $autoServiceIds = $this->configureAutoDiscoveryClients($container, $config);
+
+        $toolbar = is_bool($config['toolbar']['enabled']) ? $config['toolbar']['enabled'] : $container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug');
```
Comment (joelwurtz) [inline DependencyInjection/HttplugExtension.php]: Can we not directly set the default value of this parameter to %kernel.debug% in the configuration class ?

File: ClientFactory/PluginClientFactory.php (hunk 1 above)
Comment (Nyholm) [inline ClientFactory/PluginClientFactory.php]: Thank you

File: DependencyInjection/HttplugExtension.php [hunk 3]
```diff
@@ -212,46 +198,34 @@ private function configureAuthentication(ContainerBuilder $container, array $con
...
+            ->addArgument([])
+        ;
 
-                // tell the plugin journal what plugins we used
-                $container->getDefinition('httplug.collector.plugin_journal')
-                    ->addMethodCall('setPlugins', [$name, $arguments['plugins']]);
-            }
-        }
+        // Tell the plugin journal what plugins we used
+        $container->getDefinition('httplug.collector.plugin_journal')
```
Comment (sagikazarmark) [inline DependencyInjection/HttplugExtension.php]: IMO this is still toolbar dependent. Service only declared in data_collector.xml?

File: DependencyInjection/ProfilerExtension.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * This class is responsible for loading profiler tools and the toolbar. This extension should generally not be used
+ * in production.
+ *
+ * @author David Buchmann <mail@davidbu.ch>
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class ProfilerExtension
```
Comment (sagikazarmark) [inline DependencyInjection/ProfilerExtension.php]: I don't see the reason for a separate class. Separate methods are oka and enough IMO

File: Resources/config/services.xml [hunk 5]
```diff
@@ -8,6 +8,8 @@
             <tag name="kernel.event_subscriber"/>
         </service>
 
+        <service id="httplug.collector.plugin_journal" class="Http\HttplugBundle\Collector\PluginJournal" public="false" />
```
Comment (sagikazarmark) [inline Resources/config/services.xml]: Why is this moved here?

Comment (Nyholm) [inline Resources/config/services.xml]: It is moved here because I did not want to send a list of Plugins to the ProfilerExtension. The overhead of having it here are minimum and I thought the code would be cleaner. 

It also explains https://github.com/php-http/HttplugBundle/pull/97#discussion_r71820163

File: DependencyInjection/ProfilerExtension.php (hunk 4 above)
Comment (Nyholm) [inline DependencyInjection/ProfilerExtension.php]: The HttplugExtension was getting large enough. That's why I added a new class. No other real reasons.

Comment (sagikazarmark) [inline DependencyInjection/ProfilerExtension.php]: I still think it would be better to have all the code in one class, even if it is large. IMO it is easier to read.

File: Resources/config/services.xml (hunk 5 above)
Comment (sagikazarmark) [inline Resources/config/services.xml]: I would be happier if we had it completely separated. I don't think it makes the code cleaner.

Comment (Nyholm) [inline Resources/config/services.xml]: I'll try to make that change tomorrow.

Comment (sagikazarmark) [inline Resources/config/services.xml]: I already did, see #100

Comment (Nyholm) [inline Resources/config/services.xml]: I just notice that. Awesome. Thanks. I'll review it tomorrow.

**Comment by Nyholm** [comment]
👍 I like this. This could be merged into master.

**Comment by sagikazarmark** [comment]
> All clients are registered with the PluginClient. (even in production)

I would add this to the changelog
